test/x509_internal_test.c

```diff
@@ -58,22 +58,92 @@ static IP_TESTDATA a2i_ipaddress_tests[] = {
     {"127.0.0.1", "\x7f\x00\x00\x01", 4},
     {"1.2.3.4", "\x01\x02\x03\x04", 4},
     {"1.2.3.255", "\x01\x02\x03\xff", 4},
-    {"1.2.3", NULL, 0},
-    {"1.2.3 .4", NULL, 0},
+    {"255.255.255.255", "\xff\xff\xff\xff", 4},
 
+    {"::", "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", 16},
     {"::1", "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01", 16},
+    {"::01", "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01", 16},
+    {"::0001", "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01", 16},
+    {"ffff::", "\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", 16},
+    {"ffff::1", "\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01", 16},
+    {"1::2", "\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02", 16},
     {"1:1:1:1:1:1:1:1", "\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01", 16},
     {"2001:db8::ff00:42:8329", "\x20\x01\x0d\xb8\x00\x00\x00\x00\x00\x00\xff\x00\x00\x42\x83\x29", 16},
+    {"::1.2.3.4", "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x02\x03\x04", 16},
+    {"ffff:ffff:ffff:ffff:ffff:ffff:1.2.3.4", "\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x01\x02\x03\x04", 16},
+
     {"1:1:1:1:1:1:1:1.test", NULL, 0},
     {":::1", NULL, 0},
     {"2001::123g", NULL, 0},
 
-    {"example.test", NULL, 0},
+    /* Too few IPv4 components. */
+    {"1", NULL, 0 },
+    {"1.", NULL, 0 },
+    {"1.2", NULL, 0 },
+    {"1.2.", NULL, 0 },
+    {"1.2.3", NULL, 0 },
+    {"1.2.3.", NULL, 0 },
```

Comment:
Some of those are valid IPv4 addresses:
- "1" is "0.0.0.1"
- "1.2" is "1.0.0.2" (and so 127.1 is 127.0.0.1)
- "1.2.3" is "1.2.0.3"

I'm not saying we should accept them, but they are valid and supported by almost all tools.

Reply:
This is a valid point. The current implementation in openssl 3.3 considers those IP addresses as invalid. I'm not sure if we should change the behavior in this PR. I think this needs some more thought. I prefer to keep behavior same. The code in question is part of x509 subsystem. The reason I want to keep 1.2 input refused is it may be source of confusion. Consider situation as follows:
   - the address entered as 1.2 in ASCIIZ form
   -  it gets converted to 1.0.0.2 binary and encoded using ber/der
   - converting the binary value back to ASCIIZ will give 1.0.0.2 which is different to what user entered.

Reply:
+1 to what @Sashan said. If OpenSSL wants to switch to accepting those syntaxes, I think that's best done separately. I wrote those tests because test coverage was generally lacking.

That said, what "is" a valid IPv4 address is very context-dependent. If you ask the IETF URL RFC, the answer is [no](https://www.rfc-editor.org/rfc/rfc3986.html#page-20).

But if you ask the WHATWG URL standard, those are allowed, but then you're also allowed to write components in [hex or octal](https://url.spec.whatwg.org/#concept-ipv4-parser), which OpenSSL has also never supported.

If you ask [RFC 952](https://www.rfc-editor.org/rfc/rfc952), cited by [RFC 1123](https://www.rfc-editor.org/rfc/rfc1123#page-13), you have to have four decimal numbers, so the answer is again no.

Given OpenSSL's previous behavior, the more narrow syntax seems to be the best path here.